### PR TITLE
feat(communities): Update permissions list for spectated community tokens model

### PR DESCRIPTION
**Pull request by borismelnik**
Fixes: #11480

### What does the PR do

Adds method to update spectated community token model by shared addresses

### Affected areas

Communities

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [ ] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [ ] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)

- [ ] I've checked the design and this PR matches it

### CODE REVIEW

File: src/app/modules/main/communities/controller.nim [hunk 1]
```diff
@@ -162,6 +162,11 @@ proc init*(self: Controller) =
     let args = CommunityTokenMetadataArgs(e)
     self.delegate.onCommunityTokenMetadataAdded(args.communityId, args.tokenMetadata)
 
+  self.events.on(SIGNAL_CHECK_PERMISSIONS_TO_JOIN_RESPONSE) do(e: Args):
+    let args = CheckPermissionsToJoinResponseArgs(e)
+    if (args.communityId == self.tmpCommunityId):
+      self.delegate.onCommunityCheckPermissionsToJoinResponse(args.communityId, self.tmpAddressesToShare ,args.checkPermissionsToJoinResponse)
```
Comment (jrainville) [inline src/app/modules/main/communities/controller.nim:168]:
```suggestion
      self.delegate.onCommunityCheckPermissionsToJoinResponse(args.communityId, self.tmpAddressesToShare, args.checkPermissionsToJoinResponse)
```

File: src/app/modules/main/communities/module.nim [hunk 2]
```diff
@@ -44,6 +44,7 @@ type
     moduleLoaded: bool
     curatedCommunitiesLoaded: bool
     communityTokensModule: community_tokens_module.AccessInterface
+    tmpCommunityId: string
```
Comment (jrainville) [inline src/app/modules/main/communities/module.nim:47]: This one is not used

File: src/app/modules/main/communities/controller.nim [hunk 3]
```diff
@@ -398,3 +403,8 @@ proc authenticateWithCallback*(self: Controller) =
 
 proc getCommunityPublicKeyFromPrivateKey*(self: Controller, communityPrivateKey: string): string =
   result = self.communityService.getCommunityPublicKeyFromPrivateKey(communityPrivateKey)
+
+proc asyncCheckPermissionsToJoin*(self: Controller, communityId: string, addressesToShare: seq[string]) =
+  self.tmpCommunityId = communityId
+  self.tmpAddressesToShare = addressesToShare
+  self.communityService.asyncCheckPermissionsToJoin(communityId)
```
Comment (jrainville) [inline src/app/modules/main/communities/controller.nim:410]: You're not passing `addressesToShare` here, so you're just checking the basic addresses

File: src/app/modules/main/communities/module.nim [hunk 4]
```diff
@@ -513,4 +514,45 @@ method prepareTokenModelForCommunity*(self: Module, communityId: string) =
...
     tokenPermissionsItems.add(tokenPermissionItem)
 
-  self.view.spectatedCommunityPermissionModel.setItems(tokenPermissionsItems)
\ No newline at end of file
+  self.view.spectatedCommunityPermissionModel.setItems(tokenPermissionsItems)
+
+method updateTokenModelForCommunity*(self: Module, communityId: string, sharedAddresses: seq[string]) =
+  self.controller.asyncCheckPermissionsToJoin(communityId, sharedAddresses)
+
+method onCommunityCheckPermissionsToJoinResponse*(self: Module, communityId: string, sharedAddresses: seq[string], checkPermissionsToJoinResponse: CheckPermissionsToJoinResponseDto) =
```
Comment (jrainville) [inline src/app/modules/main/communities/module.nim:522]: I don't think you need `sharedAddresses`. It should be passed to the service to do the checks and here you don,t use them (you don't need to keep them as tmp in that case)

**Review by jrainville — CHANGES_REQUESTED** [review]
This is not using the shared addresses to check the permissions

File: src/app/modules/main/communities/module.nim (hunk 2 above)
Comment (jrainville) [reply src/app/modules/main/communities/module.nim:47]: :scissors:

**Review by jrainville — APPROVED** [review]

**Comment by borismelnik** [comment]
> Hi @borismelnik , the table does update with the communities current permissions but there is some unexpected behaviour during use.
> 
>  2023-08-08_15-22-51.mp4 
> 
> 1. OPEN Delay in evaluating permissions
>    It took around 5 seconds for the permission to be recognised as met. The user will not be aware that it's taking time to be evaluated. Are there any concerns about this taking even longer due to network conditions or a large number of wallets on the account?
> 
> 2. OPEN The table is not updated after unselecting the wallet that meets the condition
>    After unselecting the second wallet, the permissions table did not update to show that the permission is no longer met by the remaining selected wallet.
> 
> 3. OPEN The table is showing a permission without a value
>    After setting a permission for "view", a permission is shown for "view and post" even though none is configured. This also happens the other way around.
> 
> 4. OPEN The selected wallets and airdrop do not persist
>    After selecting a wallet to share and airdrop, reopening the "select addresses to join" dialog shows the default selection of all wallets.

Hey! As I understand, that issue just about the model implementation (backend) and the UI will not work yet. @jrainville am I correct?

**Comment by jrainville** [comment]
> Then hook the new model to the UI for the join part (no need to do the edit hook for now). The components are all in the storybook and might already be partially hooked at that point.

Sadly that is not correct.

I wrote in that issue that we should hook it up to the UI:

> Then hook the new model to the UI for the join part (no need to do the edit hook for now). The components are all in the storybook and might already be partially hooked at that point.

> Acceptance criteria
> That model is hooked to the join community popup (CommunityIntroDialog)

At least for the Join flow (edit was meant to be done later. The edit modal part is to be done in this issue: https://github.com/status-im/status-desktop/issues/11822

**Comment by caybro** [comment]
> > Acceptance criteria
> > That model is hooked to the join community popup (CommunityIntroDialog)
> 
> At least for the Join flow (edit was meant to be done later. The edit modal part is to be done in this issue: #11822

The UI part of the Edit flow has been merged already as well in this PR: https://github.com/status-im/status-desktop/pull/11765

**Comment by jrainville** [comment]
@glitchminer I finished hooking up the join form to the temporary model.

Now, it will update correctly when you open, check and uncheck accounts.

Two caveats:
1. OpenSea is a bit capricious and sometimes throws out `unsuccessful request: 429 Too Many Requests`. Dario is looking at it.
2. Sometimes it's slow to get the update. We'll probably want some kind of loading indicator. @caybro do we already have that? Either way, we can do it in another issue.

[permissions.webm](https://github.com/status-im/status-desktop/assets/11926403/d5c5bcfd-3006-41c4-bd5c-a6341f4212e4)

**Comment by caybro** [comment]
> 2. Sometimes it's slow to get the update. We'll probably want some kind of loading indicator. @caybro do we already have that? Either way, we can do it in another issue.

No we don't :/ It's really just the UI recalculating the bindings and everything (perhaps some of the tasks could be offloaded to the backend)

**Comment by jrainville** [comment]
I added an issue here to add a basic loading indicator. We can have designs for it later. https://github.com/status-im/status-desktop/issues/11893

**Comment by jrainville** [comment]
status-go is merged. I rebased.

@glitchminer you can test. The issue with OpenSea turns out to be more problematic than I thought. They just shut down support for Goerli. 

My suggestion is to test this but with ERC20s instead.

**Comment by glitchminer** [comment]
Adding item 5.

Permissions are not evaluated correctly when multiple are combined. For example, 1 for join community and a 2nd for view and post.

**Comment by jrainville** [comment]
Alright, turns out the channel permissions were not hooked properly. We didn't get the chats correctly and we just didn't clal the check permission function for channels.

This is all fixed in the last two PRs.

I'll request review from everyone.

@glitchminer I think you can still start testing. Again, only with ETH for now since custom ERC20s are broken and collectibles don't work on testnet because of OpenSea.

**Comment by jrainville** [comment]
Also, new status-go PR: https://github.com/status-im/status-go/pull/3910

**Comment by osmaczko** [comment]
> I added an issue here to add a basic loading indicator. We can have designs for it later. #11893

I added naive version for that for different view, I believe you can reuse the property exposed there: https://github.com/status-im/status-desktop/pull/11818

https://github.com/status-im/status-desktop/assets/33099791/b11201be-abd3-4f6d-be64-d5b69c1790b0

File: src/app/modules/main/communities/module.nim [hunk 5]
```diff
@@ -531,13 +531,71 @@ method callbackFromAuthentication*(self: Module, authenticated: bool) =
...
+  self.view.spectatedCommunityPermissionModel.setItems(tokenPermissionsItems)
+  self.updateTokenModelForCommunity(communityId, @[])
+
+proc applyPermissionResponse*(self: Module, communityId: string, permissions: Table[string, CheckPermissionsResultDto]) =
+  let community = self.controller.getCommunityById(communityId)
+  for id, criteriaResult in permissions:
+    if community.tokenPermissions.hasKey(id):
+      let tokenPermissionItem = self.view.spectatedCommunityPermissionModel.getItemById(id)
+      if tokenPermissionItem.id == "":
+        continue
```
Comment (osmaczko) [inline src/app/modules/main/communities/module.nim:556]:
```suggestion
      if tokenPermissionItem.id == "":
        warn "no permission in model", id
        continue
```

`spectatedCommunityPermissionModel` should already contain all permissions from response, right?

File: src/app/modules/main/communities/module.nim [hunk 6]
```diff
@@ -531,13 +531,71 @@ method callbackFromAuthentication*(self: Module, authenticated: bool) =
...
+          tokenCriteriaItem.symbol,
+          tokenCriteriaItem.name,
+          tokenCriteriaItem.amount,
+          tokenCriteriaItem.`type`,
+          tokenCriteriaItem.ensPattern,
+          criteriaResult.criteria[index]
+        )
+
+        if criteriaResult.criteria[index] == false:
+          permissionSatisfied = false
```
Comment (osmaczko) [inline src/app/modules/main/communities/module.nim:573]: This seems wrong. Permission is satisfied if any criteria is met :thinking: Have you tested a permission with more than one criteria?

File: src/app/modules/main/communities/module.nim [hunk 7]
```diff
@@ -531,13 +531,71 @@ method callbackFromAuthentication*(self: Module, authenticated: bool) =
...
+
+        updatedTokenCriteriaItems.add(updatedTokenCriteriaItem)
+
+      let updatedTokenPermissionItem = initTokenPermissionItem(
+          tokenPermissionItem.id,
+          tokenPermissionItem.`type`,
+          updatedTokenCriteriaItems,
```
Comment (osmaczko) [inline src/app/modules/main/communities/module.nim:580]: AFAICS We don't need `updatedTokenCriteriaItems` container.

```suggestion
          tokenPermissionItem.getTokenCriteria().getItems(),
```

File: ui/app/AppLayouts/Chat/stores/RootStore.qml [hunk 8]
```diff
@@ -56,6 +56,13 @@ QtObject {
         }
     }
 
+    function prepareTokenModelForCommunity(publicKey) {
+        root.communitiesModuleInst.prepareTokenModelForCommunity(publicKey)
+    }
+
+    property var permissionsModel: !!root.communitiesModuleInst.spectatedCommunityPermissionModel ?
```
Comment (osmaczko) [inline ui/app/AppLayouts/Chat/stores/RootStore.qml:63]:
```suggestion
    readonly property var permissionsModel: !!root.communitiesModuleInst.spectatedCommunityPermissionModel ?
```

File: src/app/modules/main/chat_section/module.nim [hunk 9]
```diff
@@ -769,7 +769,8 @@ method onCommunityTokenPermissionDeleted*(self: Module, communityId: string, per
   singletonInstance.globalEvents.showCommunityTokenPermissionDeletedNotification(communityId, "Community permission deleted", "A token permission has been removed")
 
 method onCommunityTokenPermissionCreated*(self: Module, communityId: string, tokenPermission: CommunityTokenPermissionDto) =
-  let chats = self.controller.getChatDetailsByIds(tokenPermission.chatIDs)
+  let community = self.controller.getCommunityById(communityId)
+  let chats = community.getCommunityChats(tokenPermission.chatIds)
```
[the remark this answers is not in the document]
Comment (jrainville) [reply src/app/modules/main/chat_section/module.nim:773]: I think it's the classic case of community channels having a different ID than a personal chat. Or maybe the community chats are no longer in the chat cache? 

Anyway, it's better to go directly to the source of the community chats. In most cases, I need to have the CommunityDto anyway, so one less call to the services.

File: src/app/modules/main/communities/module.nim (hunk 6 above)
Comment (jrainville) [reply src/app/modules/main/communities/module.nim:573]: This code was just copied by Boris from the chat_section module to the communities module, so I'm not 100% sure of what everything does.

I think this is right, criterias act as logical ANDs. Having multiple permissions acts as `OR`.
(I had to check the design as well to refresh my memory :smile: https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?type=design&node-id=2367-280302&mode=design&t=KT4mjhyWRIJbGmUs-0)

File: src/app/modules/main/communities/module.nim (hunk 5 above)
Comment (jrainville) [reply src/app/modules/main/communities/module.nim:556]: It should yes

File: src/app/modules/main/communities/module.nim (hunk 7 above)
Comment (jrainville) [reply src/app/modules/main/communities/module.nim:580]: I think it's needed to pass the new criteria result
```nim
initTokenCriteriaItem(
  ...
  criteriaResult.criteria[index]
)
```

File: src/app/modules/main/communities/module.nim (hunk 6 above)
Comment (osmaczko) [reply src/app/modules/main/communities/module.nim]: Ah, right, sorry for confusion :pray:

**Review by osmaczko — APPROVED** [review]

**Comment by jrainville** [comment]
I'm gonna merge now. We can test it thoroughly in the RC.

**Pull request merged**